### PR TITLE
JDK-8319626: Override toString() for ZipFile

### PR DESCRIPTION
**Pull request by justin-curtis-lu**
Please review this PR and [CSR](https://bugs.openjdk.org/browse/JDK-8319982) which overrides and provides an implementation of `toString()` in _java.util.zip.ZipFile_ (and by extension, _java.util.jar.JarFile_).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8319982](https://bugs.openjdk.org/browse/JDK-8319982) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8319626](https://bugs.openjdk.org/browse/JDK-8319626): Override toString() for ZipFile (**Enhancement** - P4)
 * [JDK-8319982](https://bugs.openjdk.org/browse/JDK-8319982): Override toString() for ZipFile (**CSR**)

### Reviewers
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)
 * [Sean Coffey](https://openjdk.org/census#coffeys) (@coffeys - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16643/head:pull/16643` \
`$ git checkout pull/16643`

Update a local copy of the PR: \
`$ git checkout pull/16643` \
`$ git pull https://git.openjdk.org/jdk.git pull/16643/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16643`

View PR using the GUI difftool: \
`$ git pr show -t 16643`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16643.diff">https://git.openjdk.org/jdk/pull/16643.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16643#issuecomment-1809211899)

### CODE REVIEW

File: src/java.base/share/classes/java/util/zip/ZipFile.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/ZipFile.java:498]: I don't think this is too helpful as the file name may not be unique or there may be several instances of ZipFile that are backed by the same zip file. Can you try `"" + file + Integer.toHexString(System.identityHashCode(file))` ?

Comment (justin-curtis-lu) [reply src/java.base/share/classes/java/util/zip/ZipFile.java:493]: Hi Alan,

Thanks for taking a look. I updated the toString() value to the one you suggested, and also dropped the specific aspects of the specification. 

I am not sure if you have a preference one way or another regarding providing the full path versus just the file name, but I can switch the full path for just the file name if need be.

File: src/java.base/share/classes/java/util/zip/ZipFile.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/ZipFile.java:494]: This will give you a String like `foo/bar.zip174d20a`, I think you'll need some separator and "@" is often used with the hex representation of the identity hash. Using `name + "@" + Objects.toIdentityString(this);` is another option to consider.

Comment (pavelrappo) [reply src/java.base/share/classes/java/util/zip/ZipFile.java:494]: @justin-curtis-lu, for inspiration and consistency, you might want to have a look at this implementation: https://github.com/openjdk/jdk/blob/1d9688667e667dc710d64e52f1e918e047beaca3/src/java.base/share/classes/java/util/concurrent/CompletableFuture.java#L2605-L2628

Comment (pavelrappo) [reply src/java.base/share/classes/java/util/zip/ZipFile.java:494]: Generally, search codebase for `super.toString()` to get some idea as to how people incorporate superclass `toString()` into subclass one.

**Comment by ecki** [comment]
If it does not return a useable string (I.e. not the entry name) then the Javadoc does not have to be prescriptive about the format to expect, but it should make it very clear it’s only a debug string. People are just too used to path.toString() being correct.

File: src/java.base/share/classes/java/util/zip/ZipFile.java
Comment (justin-curtis-lu) [reply src/java.base/share/classes/java/util/zip/ZipFile.java:494]: Thanks Pavel and Alan. 

I updated the return value to be separated by '@'. I also updated the specification wording so that users will be cautious when using this value as @ecki suggested. I used phrasing similar to the second example in Effective Java Item 12. I expect the 'subject to change' will ensure people do not depend on it.

File: src/java.base/share/classes/java/util/zip/ZipFile.java
Comment (ecki) [reply src/java.base/share/classes/java/util/zip/ZipFile.java:497]: I like the new wording (have no oppinion if absolute path is better).

Comment (jaikiran) [reply src/java.base/share/classes/java/util/zip/ZipFile.java:497]: Hello Justin,

> I am not sure if you have a preference one way or another regarding providing the full path versus just the file name, but I can switch the full path for just the file name if need be.

My opinion is that we should not use the absolute path here. Section 2.1 of secure coding guidelines https://www.oracle.com/java/technologies/javase/seccodeguide.html#2-1 suggests not to include full paths in exception messages.

With the proposed change to the toString() method here, which uses absolute paths, I think it would then mean that we would have to review (within the JDK) usages of (explicit or implicit) `ZipFile.toString()` to prevent accidentally including the complete paths in the exceptions, like in the case below:

```java
final ZipFile zf = new ZipFile("/home/me/xyz.zip");
... // do something 
throw new Exception("failed to handle zipfile " + zf);
```

Comment (justin-curtis-lu) [reply src/java.base/share/classes/java/util/zip/ZipFile.java:497]: Thanks Jai, that makes sense. Replaced full path with just the base name in latest commit.

Comment (AlanBateman) [reply src/java.base/share/classes/java/util/zip/ZipFile.java:497]: I think second paragraph of the method description needs to be dropped. Documenting the representation and then saying it is subject to change is going to cause problems now the road.

Comment (LanceAndersen) [reply src/java.base/share/classes/java/util/zip/ZipFile.java:497]: I agree we should probably remove the 2nd paragraph

File: src/java.base/share/classes/java/util/zip/ZipFile.java
Comment (jaikiran) [inline src/java.base/share/classes/java/util/zip/ZipFile.java:494]: Hello Justin, relying on `res.zsrc.key.file` to get to the file name can cause troubles. For example, consider this usage (which I ran in jshell) with this proposed change in this PR:

```console
jshell> import java.util.zip.*

jshell> ZipFile zf = new ZipFile("/tmp/workdir.zip")
zf ==> workdir.zip@34c45dca

jshell> zf.close()

jshell> zf.toString()
|  Exception java.lang.NullPointerException: Cannot read field "key" because "this.res.zsrc" is null
|        at ZipFile.toString (ZipFile.java:494)
|        at (#4:1)


```
What I do here is that I call `ZipFile.toString()` after I (intentionally) close the `ZipFile`. The `toString()` call leads to `NullPointerException` because on closing the `ZipFile` instance we clean up the resource it holds on to.

I think what we can do here is, something like (the following diff was generated on top of the current state of this PR):

```diff
diff --git a/src/java.base/share/classes/java/util/zip/ZipFile.java b/src/java.base/share/classes/java/util/zip/ZipFile.java
index 67a5e65089f..2b6996294e1 100644
--- a/src/java.base/share/classes/java/util/zip/ZipFile.java
+++ b/src/java.base/share/classes/java/util/zip/ZipFile.java
@@ -95,7 +95,8 @@
  */
 public class ZipFile implements ZipConstants, Closeable {
 
-    private final String name;     // zip file name
+    private final String filePath;     // zip file path
+    private final String fileName;     // name of the file
     private volatile boolean closeRequested;
 
     // The "resource" used by this zip file that needs to be
@@ -245,7 +246,8 @@ public ZipFile(File file, int mode, Charset charset) throws IOException
         }
         Objects.requireNonNull(charset, "charset");
 
-        this.name = name;
+        this.filePath = name;
+        this.fileName = file.getName();
         long t0 = System.nanoTime();
 
         this.res = new CleanableResource(this, ZipCoder.get(charset), file, mode);
@@ -483,7 +485,7 @@ public int available() throws IOException {
      * @return the path name of the ZIP file
      */
     public String getName() {
-        return name;
+        return filePath;
     }
 
     /**
@@ -491,7 +493,7 @@ public String getName() {
      */
     @Override
     public String toString() {
-        return res.zsrc.key.file.getName()
+        return this.fileName
                 + "@" + Integer.toHexString(System.identityHashCode(this));
     }
 


```
The above example usage now works even after I close the ZipFile. I haven't run other tests, but given that this doesn't change any APIs or their implementation, I think the tests would run fine too.

Do you see any issues with this?

Comment (justin-curtis-lu) [reply src/java.base/share/classes/java/util/zip/ZipFile.java]: Hi Jai,

Thank you for the detailed response and suggested alternative.

My intentions were to get the file name by leveraging the existing code without introducing an additional field if possible, but I did not realize the `NPE` it would cause when closing the ZipFile as you demonstrated.

I appreciate your guidance, updated the PR to reflect your suggestions. I don't see how this would cause any issues with tests as you stated, but checked tiers 1-3 and all is good there.

Comment (jaikiran) [reply src/java.base/share/classes/java/util/zip/ZipFile.java]: Hello Justin, thank you for doing the update and running the tests. This latest state of the PR (commit `4ba2bd47`) looks good to me. I have no other review comments.

Please wait for an additional review from someone else before integrating.

**Comment by eirbjo** [comment]
I'm seeing the `ZipSourceCache` fail on GHA on `windows-x64`:

```
FAILED     ZipSourceCache::testKeySourceMapping 'testKeySourceMapping()'
java.nio.file.FileSystemException: 1702471080605-bug8317678.zip: The process cannot access the file because it is being used by another process
	at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:92)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
	at java.base/sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:273)
	at java.base/sun.nio.fs.AbstractFileSystemProvider.deleteIfExists(AbstractFileSystemProvider.java:109)
	at java.base/java.nio.file.Files.deleteIfExists(Files.java:1191)
	at ZipSourceCache.cleanup(ZipSourceCache.java:65)
```

Build summary: https://github.com/eirbjo/jdk/actions/runs/7194395508

**Comment by justin-curtis-lu** [comment]
> I'm seeing the `ZipSourceCache` fail on GHA on `windows-x64`:

Hi Eirik,

Please let me know if I'm misunderstanding,

But this change hasn't been integrated into the master branch yet. Your linked PR does not appear to contain the contents of this change either. So I would expect that GHA error is unrelated to this change.  I have re-ran the test from the GHA failure (against this PR's contents) on various platforms and all appear to pass for me.

**Comment by eirbjo** [comment]
> > I'm seeing the `ZipSourceCache` fail on GHA on `windows-x64`:
> 
> Hi Eirik,
> 
> Please let me know if I'm misunderstanding,

I claim `TooManyOpenTabsException` :-) Sorry, this comment was for #16115

**Review by AlanBateman — APPROVED** [review]
The description looks fine, there is similar "for debugging" in several other APIs.

**Review by coffeys — APPROVED** [review]

**Pull request closed without merging**
